From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: unifi-qemu maintainers <qemu@localhost>
Date: Wed, 16 Sep 2026 21:30:00 +0200
Subject: [PATCH] hw/i386: expose ACPI header identity defaults

Add opt-in x86 machine properties for the remaining ACPI table header
identity fields used by the malware-analysis profile: OEM Revision,
Creator ID, and Creator Revision.

The defaults stay identical to upstream QEMU.  The launch layer only passes
these properties when the analysis profile is enabled, so ordinary Q35
machines from the same binary keep their normal ACPI identity.
---
 hw/acpi/aml-build.c         | 26 ++++++++++++++--
 hw/i386/acpi-build.c        |  2 ++
 hw/i386/x86.c               | 84 +++++++++++++++++++++++++++++++++++++++++++++
 include/hw/acpi/aml-build.h |  3 ++
 include/hw/i386/x86.h       |  6 ++++
 5 files changed, 118 insertions(+), 3 deletions(-)

diff --git a/hw/acpi/aml-build.c b/hw/acpi/aml-build.c
index 2d5826a8f1..12e21ccde9 100644
--- a/hw/acpi/aml-build.c
+++ b/hw/acpi/aml-build.c
@@ -1703,6 +1703,28 @@ Aml *aml_object_type(Aml *object)
     return var;
 }
 
+static uint32_t acpi_build_oem_revision = 1;
+static char acpi_build_creator_id[4] = {
+    ACPI_BUILD_APPNAME8[0],
+    ACPI_BUILD_APPNAME8[1],
+    ACPI_BUILD_APPNAME8[2],
+    ACPI_BUILD_APPNAME8[3],
+};
+static uint32_t acpi_build_creator_revision = 1;
+
+void acpi_table_set_header_defaults(uint32_t oem_revision,
+                                    const char *creator_id,
+                                    uint32_t creator_revision)
+{
+    acpi_build_oem_revision = oem_revision;
+    memset(acpi_build_creator_id, 0, sizeof(acpi_build_creator_id));
+    if (creator_id) {
+        memcpy(acpi_build_creator_id, creator_id,
+               MIN(strlen(creator_id), sizeof(acpi_build_creator_id)));
+    }
+    acpi_build_creator_revision = creator_revision;
+}
+
 void acpi_table_begin(AcpiTable *desc, GArray *array)
 {
 
@@ -1725,9 +1747,9 @@ void acpi_table_begin(AcpiTable *desc, GArray *array)
     build_append_padded_str(array, desc->oem_id, 6, '\0'); /* OEMID */
     /* OEM Table ID */
     build_append_padded_str(array, desc->oem_table_id, 8, '\0');
-    build_append_int_noprefix(array, 1, 4); /* OEM Revision */
-    g_array_append_vals(array, ACPI_BUILD_APPNAME8, 4); /* Creator ID */
-    build_append_int_noprefix(array, 1, 4); /* Creator Revision */
+    build_append_int_noprefix(array, acpi_build_oem_revision, 4);
+    g_array_append_vals(array, acpi_build_creator_id, 4);
+    build_append_int_noprefix(array, acpi_build_creator_revision, 4);
 }
 
 void acpi_table_end(BIOSLinker *linker, AcpiTable *desc)
diff --git a/hw/i386/acpi-build.c b/hw/i386/acpi-build.c
index 9446a9f862..2a94b25161 100644
--- a/hw/i386/acpi-build.c
+++ b/hw/i386/acpi-build.c
@@ -1954,6 +1954,8 @@ void acpi_build(AcpiBuildTables *tables, MachineState *machine)
     } else {
         oem_table_id = x86ms->oem_table_id;
     }
+    acpi_table_set_header_defaults(x86ms->oem_revision, x86ms->creator_id,
+                                   x86ms->creator_revision);
 
     table_offsets = g_array_new(false, true /* clear */,
                                         sizeof(uint32_t));
diff --git a/hw/i386/x86.c b/hw/i386/x86.c
index f80533df1c..4843f5df67 100644
--- a/hw/i386/x86.c
+++ b/hw/i386/x86.c
@@ -309,6 +309,68 @@ static void x86_machine_set_oem_table_id(Object *obj, const char *value,
     strncpy(x86ms->oem_table_id, value, 8);
 }
 
+static void x86_machine_get_oem_revision(Object *obj, Visitor *v,
+                                         const char *name, void *opaque,
+                                         Error **errp)
+{
+    X86MachineState *x86ms = X86_MACHINE(obj);
+    uint32_t value = x86ms->oem_revision;
+
+    visit_type_uint32(v, name, &value, errp);
+}
+
+static void x86_machine_set_oem_revision(Object *obj, Visitor *v,
+                                         const char *name, void *opaque,
+                                         Error **errp)
+{
+    X86MachineState *x86ms = X86_MACHINE(obj);
+
+    visit_type_uint32(v, name, &x86ms->oem_revision, errp);
+}
+
+static char *x86_machine_get_creator_id(Object *obj, Error **errp)
+{
+    X86MachineState *x86ms = X86_MACHINE(obj);
+
+    return g_strdup(x86ms->creator_id);
+}
+
+static void x86_machine_set_creator_id(Object *obj, const char *value,
+                                       Error **errp)
+{
+    X86MachineState *x86ms = X86_MACHINE(obj);
+    size_t len = strlen(value);
+
+    if (len > 4) {
+        error_setg(errp,
+                   "User specified "X86_MACHINE_CREATOR_ID
+                   " value is bigger than 4 bytes in size");
+        return;
+    }
+
+    memset(x86ms->creator_id, 0, 4);
+    memcpy(x86ms->creator_id, value, len);
+}
+
+static void x86_machine_get_creator_revision(Object *obj, Visitor *v,
+                                             const char *name, void *opaque,
+                                             Error **errp)
+{
+    X86MachineState *x86ms = X86_MACHINE(obj);
+    uint32_t value = x86ms->creator_revision;
+
+    visit_type_uint32(v, name, &value, errp);
+}
+
+static void x86_machine_set_creator_revision(Object *obj, Visitor *v,
+                                             const char *name, void *opaque,
+                                             Error **errp)
+{
+    X86MachineState *x86ms = X86_MACHINE(obj);
+
+    visit_type_uint32(v, name, &x86ms->creator_revision, errp);
+}
+
 static void x86_machine_get_bus_lock_ratelimit(Object *obj, Visitor *v,
                                 const char *name, void *opaque, Error **errp)
 {
@@ -368,6 +430,9 @@ static void x86_machine_initfn(Object *obj)
     x86ms->pci_irq_mask = ACPI_BUILD_PCI_IRQS;
     x86ms->oem_id = g_strndup(ACPI_BUILD_APPNAME6, 6);
     x86ms->oem_table_id = g_strndup(ACPI_BUILD_APPNAME8, 8);
+    x86ms->oem_revision = 1;
+    x86ms->creator_id = g_strndup(ACPI_BUILD_APPNAME8, 4);
+    x86ms->creator_revision = 1;
     x86ms->bus_lock_ratelimit = 0;
     x86ms->above_4g_mem_start = 4 * GiB;
 }
@@ -428,6 +493,28 @@ static void x86_machine_class_init(ObjectClass *oc, const void *data)
                                           "in ACPI table header."
                                           "The string may be up to 8 bytes in size");
 
+    object_class_property_add(oc, X86_MACHINE_OEM_REVISION, "uint32",
+                              x86_machine_get_oem_revision,
+                              x86_machine_set_oem_revision, NULL, NULL);
+    object_class_property_set_description(oc, X86_MACHINE_OEM_REVISION,
+                                          "Override the default value of field OEM Revision "
+                                          "in ACPI table header.");
+
+    object_class_property_add_str(oc, X86_MACHINE_CREATOR_ID,
+                                  x86_machine_get_creator_id,
+                                  x86_machine_set_creator_id);
+    object_class_property_set_description(oc, X86_MACHINE_CREATOR_ID,
+                                          "Override the default value of field Creator ID "
+                                          "in ACPI table header."
+                                          "The string may be up to 4 bytes in size");
+
+    object_class_property_add(oc, X86_MACHINE_CREATOR_REVISION, "uint32",
+                              x86_machine_get_creator_revision,
+                              x86_machine_set_creator_revision, NULL, NULL);
+    object_class_property_set_description(oc, X86_MACHINE_CREATOR_REVISION,
+                                          "Override the default value of field Creator Revision "
+                                          "in ACPI table header.");
+
     object_class_property_add(oc, X86_MACHINE_BUS_LOCK_RATELIMIT, "uint64_t",
                                 x86_machine_get_bus_lock_ratelimit,
                                 x86_machine_set_bus_lock_ratelimit, NULL, NULL);
diff --git a/include/hw/acpi/aml-build.h b/include/hw/acpi/aml-build.h
index f38e129719..928d2a1012 100644
--- a/include/hw/acpi/aml-build.h
+++ b/include/hw/acpi/aml-build.h
@@ -434,6 +434,9 @@ typedef struct AcpiTable {
  * @array: blob where the ACPI table will be composed/stored.
  */
 void acpi_table_begin(AcpiTable *desc, GArray *array);
+void acpi_table_set_header_defaults(uint32_t oem_revision,
+                                    const char *creator_id,
+                                    uint32_t creator_revision);
 
 /**
  * acpi_table_end:
diff --git a/include/hw/i386/x86.h b/include/hw/i386/x86.h
index 8755cad50a..2355ca6505 100644
--- a/include/hw/i386/x86.h
+++ b/include/hw/i386/x86.h
@@ -81,6 +81,9 @@ struct X86MachineState {
 
     char *oem_id;
     char *oem_table_id;
+    uint32_t oem_revision;
+    char *creator_id;
+    uint32_t creator_revision;
     /*
      * Address space used by IOAPIC device. All IOAPIC interrupts
      * will be translated to MSI messages in the address space.
@@ -103,6 +106,9 @@ struct X86MachineState {
 #define X86_MACHINE_PIC              "pic"
 #define X86_MACHINE_OEM_ID           "x-oem-id"
 #define X86_MACHINE_OEM_TABLE_ID     "x-oem-table-id"
+#define X86_MACHINE_OEM_REVISION     "x-oem-revision"
+#define X86_MACHINE_CREATOR_ID       "x-creator-id"
+#define X86_MACHINE_CREATOR_REVISION "x-creator-revision"
 #define X86_MACHINE_BUS_LOCK_RATELIMIT  "bus-lock-ratelimit"
 
 #define TYPE_X86_MACHINE   MACHINE_TYPE_NAME("x86")
-- 
2.45.2
